### PR TITLE
ttyd: 1.6.1 -> 1.6.3, generate blob from source

### PR DESCRIPTION
**Pull request by iblech**
###### Motivation for this change

The first commit in this pull request is a very ordinary version bump (1.6.1 to 1.6.3).

In the second, we fix a long-time issue with our ttyd package: The file `src/html.h` in the ttyd repository is generated from other source files in the `html/` subdirectory. Previously, we just used the pre-built file. With this pull request, we are generating it on our own.

I believe this to be in the Nix spirit, please correct me if I'm wrong. It also has the practical benefit that overriding this package is now easier -- before, any changes to the `html/` subdirectory wouldn't have any effect.

In order to avoid checking in a 300+ KiB `yarn.nix` file into nixpkgs (automatically generated from `yarn.lock`), this pull requests generates it on the fly. Please indicate whether this is weird, if so I'll just check in the generated `yarn.nix` as a static file.

The current maintainer @thoughtpolice shouldn't have to bear the work of maintaining my addition. Hence I'm adding myself as maintainer. :-)

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by SuperSandro2000** [comment]
Please check ofborg status before posting on discourse. I usually just skip red ofborgs entirely.

**Comment by iblech** [comment]
The status is only red because I'm not a trusted user, hence ofborg refuses to evaluate the expression. The expression is slightly unusual because it calls yarn2nix at evaluation time. :-) This aspect is what I solicited feedback about.

The alternative would be to skip this dynamical generation and just check in a static 300 KiB automatically-generated .nix file. Would also be totally fine with me, just wanted to know which route I should pursue.

File: pkgs/servers/ttyd/default.nix [hunk 1]
```diff
@@ -1,30 +1,62 @@
...
+  # The file src/html.h in the ttyd repository is generated from the frontend
+  # source in the html/ subdirectory. Instead of using this pre-built file, we
+  # generate it ourselves. We do this not only to keep everything honest, but
+  # also to facilitate easy overriding/patching of this package. Changes to the
+  # html/ subdirectory wouldn't have any effect without regenerating src/html.h.
+  preBuild =
+    let
+      yarnNixFile = runCommand "yarn.nix" {}
+        "${yarn2nix}/bin/yarn2nix --lockfile ${src}/html/yarn.lock --no-patch --builtin-fetchgit > $out";
+      yarnNix = callPackage yarnNixFile {};
```
Comment (Mic92) [inline pkgs/servers/ttyd/default.nix:32]: This does import from derivation, which we don't allow in nixpkgs.

Comment (Mic92) [reply pkgs/servers/ttyd/default.nix:32]: Hence the evaluation error.

Comment (SuperSandro2000) [inline pkgs/servers/ttyd/default.nix:32]: You can't do this in nixpkgs and so ofborg fails. So I think you need to add the lock file.

**Comment by iblech** [comment]
Thank you @Mic92 and @SuperSandro2000 for your comments! Accordingly, I removed the dynamic generation and checked in a static yarn.nix file.

File: pkgs/servers/ttyd/update.sh [hunk 2]
```diff
@@ -0,0 +1,10 @@
+#!/usr/bin/env nix-shell
+#!nix-shell -i bash -p yarn2nix curl
```
Comment (SuperSandro2000) [inline pkgs/servers/ttyd/update.sh:2]:
```suggestion
#!nix-shell -i bash -p curl gnugrep yarn2nix
```

**Review by Mic92 — CHANGES_REQUESTED** [review]
This is 300KB for a single package? I am against having this in nixpkgs. The problem is that this is also increases nixpkgs evaluation time massively.

**Comment by Mic92** [comment]
To be honest I rather have pre-compiled html than reviewing 10000 lines generate nix files.

**Comment by Mic92** [comment]
Sorry. For not reading the PR message before putting my comment below. I was asked to have a look at the evaluation error, so I did not pay attention to that.

File: pkgs/servers/ttyd/default.nix [hunk 3]
```diff
@@ -1,30 +1,60 @@
 { lib, stdenv, fetchFromGitHub
 , pkg-config, cmake, xxd
 , openssl, libwebsockets, json_c, libuv, zlib
+, python3, nodejs-slim, yarn, fixup_yarn_lock
+, callPackage
 }:
 
 with builtins;
```
Comment (SuperSandro2000) [inline pkgs/servers/ttyd/default.nix:9]:
```suggestion
```

**Comment by Mindavi** [comment]
Heh, I also bumped ttyd a while ago, so the bump could be dropped after a rebase: https://github.com/NixOS/nixpkgs/pull/123135. Didn't see this PR at the time. I also don't really like the large files in nixpkgs, so a :-1: from me on this. I think this would be better suited to be carried as an overlay if you want to do such things, rather than being checked in to nixpkgs.

**Pull request closed without merging**